eva-accession-clustering/src/main/java/uk/ac/ebi/eva/accession/clustering/batch/io/ClusteringWriter.java

```diff
@@ -276,32 +270,59 @@ private void clusterSubmittedVariants(List<? extends SubmittedVariantEntity> sub
         BulkOperations dbsnpBulkOperations = mongoTemplate.bulkOps(BulkOperations.BulkMode.UNORDERED,
                                                                    DbsnpSubmittedVariantEntity.class);
+        BulkOperations bulkHistoryOperations = mongoTemplate.bulkOps(BulkOperations.BulkMode.UNORDERED,
+                SubmittedVariantOperationEntity.class);
+        BulkOperations dbsnpBulkHistoryOperations = mongoTemplate.bulkOps(BulkOperations.BulkMode.UNORDERED,
+                DbsnpSubmittedVariantOperationEntity.class);
         long numUpdates = 0;
         long numDbsnpUpdates = 0;
         for (SubmittedVariantEntity submittedVariantEntity : submittedVariantEntities) {
             if (submittedVariantEntity.getClusteredVariantAccession() != null) {
                 // no need to update the rs of a submitted variant that already has the correct rs
                 continue;
             }
-            Query query = query(where("_id").is(submittedVariantEntity.getId()));
-            Update update = new Update();
-            update.set(RS_KEY, getClusteredVariantAccession(submittedVariantEntity));
+
+            long rsid = getClusteredVariantAccession(submittedVariantEntity);
+            // Query to update the RSid in submittedVariantEntity
+            Query rs_query = query(where("_id").is(submittedVariantEntity.getId()));
+            Update rs_update = new Update();
+            rs_update.set(RS_KEY, rsid);
+
+            // Query to create the update operation
+            SubmittedVariantOperationEntity update_operation = createUpdateOperations(submittedVariantEntity, rsid);
+
             if (isEvaSubmittedVariant(submittedVariantEntity)) {
-                bulkOperations.updateOne(query, update);
+                bulkOperations.updateOne(rs_query, rs_update);
+                bulkHistoryOperations.insert(update_operation);
                 ++numUpdates;
             } else {
-                dbsnpBulkOperations.updateOne(query, update);
+                dbsnpBulkOperations.updateOne(rs_query, rs_update);
+                dbsnpBulkHistoryOperations.insert(update_operation);
                 ++numDbsnpUpdates;
             }
         }
         if (numUpdates > 0) {
             bulkOperations.execute();
+            bulkHistoryOperations.execute();
         }
         if (numDbsnpUpdates > 0) {
             dbsnpBulkOperations.execute();
+            dbsnpBulkHistoryOperations.execute();
         }
     }
 
+    private SubmittedVariantOperationEntity createUpdateOperations(SubmittedVariantEntity submittedVariant, Long rsid) {
+        SubmittedVariantOperationEntity operation = new SubmittedVariantOperationEntity();
+        operation.fill(
+                EventType.UPDATED,
+                null,
+                rsid,
+                "Clustering submitted variant " + submittedVariant.getAccession() + " with rs" + rsid,
+                null
+        );
+        return operation;
```

Comment:
this code does not track what was the previous status. look at the method buildSubmittedOperation in this class. The part about "inactiveEntity" of that method is relevant here too.

Reply:
I've now mirrored the other function but must say I'm a bit confused about the purpose of the inactive entity in the case of an RS update. It makes sense in the case of deprecated object but less clear here.

Reply:
It's a matter of taking a snapshot before any change. this way you can easily compare the current SS document and the one in the history. (and only the rs field will change)